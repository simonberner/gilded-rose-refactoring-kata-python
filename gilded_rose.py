# -*- coding: utf-8 -*-

class GildedRose(object):

    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            item.update_item()


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def update_item(self):
        is_not_aged_brie = self.name != "Aged Brie"
        is_not_backstage_pass = self.name != "Backstage passes to a TAFKAL80ETC concert"
        is_not_sulfuras = self.name != "Sulfuras, Hand of Ragnaros"

        if is_not_aged_brie:
            if is_not_backstage_pass:
                if self.quality > 0:
                    if is_not_sulfuras:
                        self.quality = self.quality - 1
            else:
                if self.quality < 50:
                    self.quality = self.quality + 1
                    if not is_not_backstage_pass:
                        if self.sell_in < 11 and self.quality < 50:
                            self.quality = self.quality + 1
                        if self.sell_in < 6 and self.quality < 50:
                            self.quality = self.quality + 1
            # Updating sellIn when not Sulfuras
            if is_not_sulfuras:
                self.sell_in = self.sell_in - 1
            # Code smell: repeated conditional
            if self.sell_in < 0:
                # ifelse
                if is_not_backstage_pass:
                    if self.quality > 0:
                        if is_not_sulfuras:
                            self.quality = self.quality - 1
                else:
                    self.quality -= self.quality

        # it is aged_brie
        else:
            if self.quality < 50:
                self.quality = self.quality + 1
            self.sell_in = self.sell_in - 1
            if self.sell_in < 0 and self.quality < 50:
                self.quality = self.quality + 1

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)
