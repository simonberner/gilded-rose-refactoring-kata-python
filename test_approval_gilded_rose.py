from gilded_rose import Item, GildedRose
from approvaltests import verify_all_combinations, Options, DiffReporter
from approvaltests.reporters import PythonNativeReporter


# Check that sellIn and Quality degrades by 1 for a normal item
def test_update_quality():
    name = "normal item"
    sell_in = 10
    quality = 4

    verify_all_combinations(
        do_update_quality,
        [[str(name)], [int(sell_in)], [int(quality)]],
        options=Options().with_reporter(reporter=PythonNativeReporter())
    )


def do_update_quality(name, sell_in, quality):
    items = [Item(name, sell_in, quality)]
    gilded_rose = GildedRose(items)
    gilded_rose.update_quality()
    item = gilded_rose.items[0]
    return __print_item_as_string(item)


def __print_item_as_string(item):
    return item.name + ", sellIn: " + str(item.sell_in) + ", Quality: " + str(item.quality)
