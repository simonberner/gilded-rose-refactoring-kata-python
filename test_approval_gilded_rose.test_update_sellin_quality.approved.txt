args: ('normal item', -1, -1) => 'normal item, sellIn: -2, Quality: -1'
args: ('normal item', -1, 0) => 'normal item, sellIn: -2, Quality: 0'
args: ('normal item', -1, 4) => 'normal item, sellIn: -2, Quality: 2'
args: ('normal item', -1, 50) => 'normal item, sellIn: -2, Quality: 48'
args: ('normal item', -1, 51) => 'normal item, sellIn: -2, Quality: 49'
args: ('normal item', 0, -1) => 'normal item, sellIn: -1, Quality: -1'
args: ('normal item', 0, 0) => 'normal item, sellIn: -1, Quality: 0'
args: ('normal item', 0, 4) => 'normal item, sellIn: -1, Quality: 2'
args: ('normal item', 0, 50) => 'normal item, sellIn: -1, Quality: 48'
args: ('normal item', 0, 51) => 'normal item, sellIn: -1, Quality: 49'
args: ('normal item', 5, -1) => 'normal item, sellIn: 4, Quality: -1'
args: ('normal item', 5, 0) => 'normal item, sellIn: 4, Quality: 0'
args: ('normal item', 5, 4) => 'normal item, sellIn: 4, Quality: 3'
args: ('normal item', 5, 50) => 'normal item, sellIn: 4, Quality: 49'
args: ('normal item', 5, 51) => 'normal item, sellIn: 4, Quality: 50'
args: ('normal item', 6, -1) => 'normal item, sellIn: 5, Quality: -1'
args: ('normal item', 6, 0) => 'normal item, sellIn: 5, Quality: 0'
args: ('normal item', 6, 4) => 'normal item, sellIn: 5, Quality: 3'
args: ('normal item', 6, 50) => 'normal item, sellIn: 5, Quality: 49'
args: ('normal item', 6, 51) => 'normal item, sellIn: 5, Quality: 50'
args: ('normal item', 10, -1) => 'normal item, sellIn: 9, Quality: -1'
args: ('normal item', 10, 0) => 'normal item, sellIn: 9, Quality: 0'
args: ('normal item', 10, 4) => 'normal item, sellIn: 9, Quality: 3'
args: ('normal item', 10, 50) => 'normal item, sellIn: 9, Quality: 49'
args: ('normal item', 10, 51) => 'normal item, sellIn: 9, Quality: 50'
args: ('normal item', 11, -1) => 'normal item, sellIn: 10, Quality: -1'
args: ('normal item', 11, 0) => 'normal item, sellIn: 10, Quality: 0'
args: ('normal item', 11, 4) => 'normal item, sellIn: 10, Quality: 3'
args: ('normal item', 11, 50) => 'normal item, sellIn: 10, Quality: 49'
args: ('normal item', 11, 51) => 'normal item, sellIn: 10, Quality: 50'
args: ('Aged Brie', -1, -1) => 'Aged Brie, sellIn: -2, Quality: 1'
args: ('Aged Brie', -1, 0) => 'Aged Brie, sellIn: -2, Quality: 2'
args: ('Aged Brie', -1, 4) => 'Aged Brie, sellIn: -2, Quality: 6'
args: ('Aged Brie', -1, 50) => 'Aged Brie, sellIn: -2, Quality: 50'
args: ('Aged Brie', -1, 51) => 'Aged Brie, sellIn: -2, Quality: 51'
args: ('Aged Brie', 0, -1) => 'Aged Brie, sellIn: -1, Quality: 1'
args: ('Aged Brie', 0, 0) => 'Aged Brie, sellIn: -1, Quality: 2'
args: ('Aged Brie', 0, 4) => 'Aged Brie, sellIn: -1, Quality: 6'
args: ('Aged Brie', 0, 50) => 'Aged Brie, sellIn: -1, Quality: 50'
args: ('Aged Brie', 0, 51) => 'Aged Brie, sellIn: -1, Quality: 51'
args: ('Aged Brie', 5, -1) => 'Aged Brie, sellIn: 4, Quality: 0'
args: ('Aged Brie', 5, 0) => 'Aged Brie, sellIn: 4, Quality: 1'
args: ('Aged Brie', 5, 4) => 'Aged Brie, sellIn: 4, Quality: 5'
args: ('Aged Brie', 5, 50) => 'Aged Brie, sellIn: 4, Quality: 50'
args: ('Aged Brie', 5, 51) => 'Aged Brie, sellIn: 4, Quality: 51'
args: ('Aged Brie', 6, -1) => 'Aged Brie, sellIn: 5, Quality: 0'
args: ('Aged Brie', 6, 0) => 'Aged Brie, sellIn: 5, Quality: 1'
args: ('Aged Brie', 6, 4) => 'Aged Brie, sellIn: 5, Quality: 5'
args: ('Aged Brie', 6, 50) => 'Aged Brie, sellIn: 5, Quality: 50'
args: ('Aged Brie', 6, 51) => 'Aged Brie, sellIn: 5, Quality: 51'
args: ('Aged Brie', 10, -1) => 'Aged Brie, sellIn: 9, Quality: 0'
args: ('Aged Brie', 10, 0) => 'Aged Brie, sellIn: 9, Quality: 1'
args: ('Aged Brie', 10, 4) => 'Aged Brie, sellIn: 9, Quality: 5'
args: ('Aged Brie', 10, 50) => 'Aged Brie, sellIn: 9, Quality: 50'
args: ('Aged Brie', 10, 51) => 'Aged Brie, sellIn: 9, Quality: 51'
args: ('Aged Brie', 11, -1) => 'Aged Brie, sellIn: 10, Quality: 0'
args: ('Aged Brie', 11, 0) => 'Aged Brie, sellIn: 10, Quality: 1'
args: ('Aged Brie', 11, 4) => 'Aged Brie, sellIn: 10, Quality: 5'
args: ('Aged Brie', 11, 50) => 'Aged Brie, sellIn: 10, Quality: 50'
args: ('Aged Brie', 11, 51) => 'Aged Brie, sellIn: 10, Quality: 51'
args: ('Backstage passes to a TAFKAL80ETC concert', -1, -1) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: -2, Quality: 0'
args: ('Backstage passes to a TAFKAL80ETC concert', -1, 0) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: -2, Quality: 0'
args: ('Backstage passes to a TAFKAL80ETC concert', -1, 4) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: -2, Quality: 0'
args: ('Backstage passes to a TAFKAL80ETC concert', -1, 50) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: -2, Quality: 0'
args: ('Backstage passes to a TAFKAL80ETC concert', -1, 51) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: -2, Quality: 0'
args: ('Backstage passes to a TAFKAL80ETC concert', 0, -1) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: -1, Quality: 0'
args: ('Backstage passes to a TAFKAL80ETC concert', 0, 0) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: -1, Quality: 0'
args: ('Backstage passes to a TAFKAL80ETC concert', 0, 4) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: -1, Quality: 0'
args: ('Backstage passes to a TAFKAL80ETC concert', 0, 50) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: -1, Quality: 0'
args: ('Backstage passes to a TAFKAL80ETC concert', 0, 51) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: -1, Quality: 0'
args: ('Backstage passes to a TAFKAL80ETC concert', 5, -1) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 4, Quality: 2'
args: ('Backstage passes to a TAFKAL80ETC concert', 5, 0) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 4, Quality: 3'
args: ('Backstage passes to a TAFKAL80ETC concert', 5, 4) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 4, Quality: 7'
args: ('Backstage passes to a TAFKAL80ETC concert', 5, 50) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 4, Quality: 50'
args: ('Backstage passes to a TAFKAL80ETC concert', 5, 51) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 4, Quality: 51'
args: ('Backstage passes to a TAFKAL80ETC concert', 6, -1) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 5, Quality: 1'
args: ('Backstage passes to a TAFKAL80ETC concert', 6, 0) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 5, Quality: 2'
args: ('Backstage passes to a TAFKAL80ETC concert', 6, 4) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 5, Quality: 6'
args: ('Backstage passes to a TAFKAL80ETC concert', 6, 50) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 5, Quality: 50'
args: ('Backstage passes to a TAFKAL80ETC concert', 6, 51) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 5, Quality: 51'
args: ('Backstage passes to a TAFKAL80ETC concert', 10, -1) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 9, Quality: 1'
args: ('Backstage passes to a TAFKAL80ETC concert', 10, 0) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 9, Quality: 2'
args: ('Backstage passes to a TAFKAL80ETC concert', 10, 4) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 9, Quality: 6'
args: ('Backstage passes to a TAFKAL80ETC concert', 10, 50) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 9, Quality: 50'
args: ('Backstage passes to a TAFKAL80ETC concert', 10, 51) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 9, Quality: 51'
args: ('Backstage passes to a TAFKAL80ETC concert', 11, -1) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 10, Quality: 0'
args: ('Backstage passes to a TAFKAL80ETC concert', 11, 0) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 10, Quality: 1'
args: ('Backstage passes to a TAFKAL80ETC concert', 11, 4) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 10, Quality: 5'
args: ('Backstage passes to a TAFKAL80ETC concert', 11, 50) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 10, Quality: 50'
args: ('Backstage passes to a TAFKAL80ETC concert', 11, 51) => 'Backstage passes to a TAFKAL80ETC concert, sellIn: 10, Quality: 51'
args: ('Sulfuras, Hand of Ragnaros', -1, -1) => 'Sulfuras, Hand of Ragnaros, sellIn: -1, Quality: -1'
args: ('Sulfuras, Hand of Ragnaros', -1, 0) => 'Sulfuras, Hand of Ragnaros, sellIn: -1, Quality: 0'
args: ('Sulfuras, Hand of Ragnaros', -1, 4) => 'Sulfuras, Hand of Ragnaros, sellIn: -1, Quality: 4'
args: ('Sulfuras, Hand of Ragnaros', -1, 50) => 'Sulfuras, Hand of Ragnaros, sellIn: -1, Quality: 50'
args: ('Sulfuras, Hand of Ragnaros', -1, 51) => 'Sulfuras, Hand of Ragnaros, sellIn: -1, Quality: 51'
args: ('Sulfuras, Hand of Ragnaros', 0, -1) => 'Sulfuras, Hand of Ragnaros, sellIn: 0, Quality: -1'
args: ('Sulfuras, Hand of Ragnaros', 0, 0) => 'Sulfuras, Hand of Ragnaros, sellIn: 0, Quality: 0'
args: ('Sulfuras, Hand of Ragnaros', 0, 4) => 'Sulfuras, Hand of Ragnaros, sellIn: 0, Quality: 4'
args: ('Sulfuras, Hand of Ragnaros', 0, 50) => 'Sulfuras, Hand of Ragnaros, sellIn: 0, Quality: 50'
args: ('Sulfuras, Hand of Ragnaros', 0, 51) => 'Sulfuras, Hand of Ragnaros, sellIn: 0, Quality: 51'
args: ('Sulfuras, Hand of Ragnaros', 5, -1) => 'Sulfuras, Hand of Ragnaros, sellIn: 5, Quality: -1'
args: ('Sulfuras, Hand of Ragnaros', 5, 0) => 'Sulfuras, Hand of Ragnaros, sellIn: 5, Quality: 0'
args: ('Sulfuras, Hand of Ragnaros', 5, 4) => 'Sulfuras, Hand of Ragnaros, sellIn: 5, Quality: 4'
args: ('Sulfuras, Hand of Ragnaros', 5, 50) => 'Sulfuras, Hand of Ragnaros, sellIn: 5, Quality: 50'
args: ('Sulfuras, Hand of Ragnaros', 5, 51) => 'Sulfuras, Hand of Ragnaros, sellIn: 5, Quality: 51'
args: ('Sulfuras, Hand of Ragnaros', 6, -1) => 'Sulfuras, Hand of Ragnaros, sellIn: 6, Quality: -1'
args: ('Sulfuras, Hand of Ragnaros', 6, 0) => 'Sulfuras, Hand of Ragnaros, sellIn: 6, Quality: 0'
args: ('Sulfuras, Hand of Ragnaros', 6, 4) => 'Sulfuras, Hand of Ragnaros, sellIn: 6, Quality: 4'
args: ('Sulfuras, Hand of Ragnaros', 6, 50) => 'Sulfuras, Hand of Ragnaros, sellIn: 6, Quality: 50'
args: ('Sulfuras, Hand of Ragnaros', 6, 51) => 'Sulfuras, Hand of Ragnaros, sellIn: 6, Quality: 51'
args: ('Sulfuras, Hand of Ragnaros', 10, -1) => 'Sulfuras, Hand of Ragnaros, sellIn: 10, Quality: -1'
args: ('Sulfuras, Hand of Ragnaros', 10, 0) => 'Sulfuras, Hand of Ragnaros, sellIn: 10, Quality: 0'
args: ('Sulfuras, Hand of Ragnaros', 10, 4) => 'Sulfuras, Hand of Ragnaros, sellIn: 10, Quality: 4'
args: ('Sulfuras, Hand of Ragnaros', 10, 50) => 'Sulfuras, Hand of Ragnaros, sellIn: 10, Quality: 50'
args: ('Sulfuras, Hand of Ragnaros', 10, 51) => 'Sulfuras, Hand of Ragnaros, sellIn: 10, Quality: 51'
args: ('Sulfuras, Hand of Ragnaros', 11, -1) => 'Sulfuras, Hand of Ragnaros, sellIn: 11, Quality: -1'
args: ('Sulfuras, Hand of Ragnaros', 11, 0) => 'Sulfuras, Hand of Ragnaros, sellIn: 11, Quality: 0'
args: ('Sulfuras, Hand of Ragnaros', 11, 4) => 'Sulfuras, Hand of Ragnaros, sellIn: 11, Quality: 4'
args: ('Sulfuras, Hand of Ragnaros', 11, 50) => 'Sulfuras, Hand of Ragnaros, sellIn: 11, Quality: 50'
args: ('Sulfuras, Hand of Ragnaros', 11, 51) => 'Sulfuras, Hand of Ragnaros, sellIn: 11, Quality: 51'
args: ('Unicorn (conjured)', -1, -1) => 'Unicorn (conjured), sellIn: -2, Quality: -1'
args: ('Unicorn (conjured)', -1, 0) => 'Unicorn (conjured), sellIn: -2, Quality: 0'
args: ('Unicorn (conjured)', -1, 4) => 'Unicorn (conjured), sellIn: -2, Quality: 2'
args: ('Unicorn (conjured)', -1, 50) => 'Unicorn (conjured), sellIn: -2, Quality: 48'
args: ('Unicorn (conjured)', -1, 51) => 'Unicorn (conjured), sellIn: -2, Quality: 49'
args: ('Unicorn (conjured)', 0, -1) => 'Unicorn (conjured), sellIn: -1, Quality: -1'
args: ('Unicorn (conjured)', 0, 0) => 'Unicorn (conjured), sellIn: -1, Quality: 0'
args: ('Unicorn (conjured)', 0, 4) => 'Unicorn (conjured), sellIn: -1, Quality: 2'
args: ('Unicorn (conjured)', 0, 50) => 'Unicorn (conjured), sellIn: -1, Quality: 48'
args: ('Unicorn (conjured)', 0, 51) => 'Unicorn (conjured), sellIn: -1, Quality: 49'
args: ('Unicorn (conjured)', 5, -1) => 'Unicorn (conjured), sellIn: 4, Quality: -1'
args: ('Unicorn (conjured)', 5, 0) => 'Unicorn (conjured), sellIn: 4, Quality: 0'
args: ('Unicorn (conjured)', 5, 4) => 'Unicorn (conjured), sellIn: 4, Quality: 2'
args: ('Unicorn (conjured)', 5, 50) => 'Unicorn (conjured), sellIn: 4, Quality: 48'
args: ('Unicorn (conjured)', 5, 51) => 'Unicorn (conjured), sellIn: 4, Quality: 49'
args: ('Unicorn (conjured)', 6, -1) => 'Unicorn (conjured), sellIn: 5, Quality: -1'
args: ('Unicorn (conjured)', 6, 0) => 'Unicorn (conjured), sellIn: 5, Quality: 0'
args: ('Unicorn (conjured)', 6, 4) => 'Unicorn (conjured), sellIn: 5, Quality: 2'
args: ('Unicorn (conjured)', 6, 50) => 'Unicorn (conjured), sellIn: 5, Quality: 48'
args: ('Unicorn (conjured)', 6, 51) => 'Unicorn (conjured), sellIn: 5, Quality: 49'
args: ('Unicorn (conjured)', 10, -1) => 'Unicorn (conjured), sellIn: 9, Quality: -1'
args: ('Unicorn (conjured)', 10, 0) => 'Unicorn (conjured), sellIn: 9, Quality: 0'
args: ('Unicorn (conjured)', 10, 4) => 'Unicorn (conjured), sellIn: 9, Quality: 2'
args: ('Unicorn (conjured)', 10, 50) => 'Unicorn (conjured), sellIn: 9, Quality: 48'
args: ('Unicorn (conjured)', 10, 51) => 'Unicorn (conjured), sellIn: 9, Quality: 49'
args: ('Unicorn (conjured)', 11, -1) => 'Unicorn (conjured), sellIn: 10, Quality: -1'
args: ('Unicorn (conjured)', 11, 0) => 'Unicorn (conjured), sellIn: 10, Quality: 0'
args: ('Unicorn (conjured)', 11, 4) => 'Unicorn (conjured), sellIn: 10, Quality: 2'
args: ('Unicorn (conjured)', 11, 50) => 'Unicorn (conjured), sellIn: 10, Quality: 48'
args: ('Unicorn (conjured)', 11, 51) => 'Unicorn (conjured), sellIn: 10, Quality: 49'
